# **Mercado de casas para alquiler en USA**
---

## **Contexto comercial y objetivo comercial**







El alquiler de casas es un mecanismo moderno para lograr obtener  un alojamiento del tipo temporal. Ante la creciente imposibilidad de aspirar hacia la obtención de una casa propia, el mercado de alquileres se consolidó como una de las soluciones más rápidas para resolver este problema y por lo cual, con el crecimiento poblacional vemos que es necesario generar cada vez más información para las personas que están entrando en esa etapa de su vida, en la cual aspiran hacia la independencia de su hogar, en criollo, "Abandonar el nido".A su vez cabe recalcar que cada vez mas se esta generando movimiento entre diferentes locaciones para llevar a cabo trabajos tanto presencial como con la modalidad homeoffice la cual a partir de la pandemia de Covid-19 se vio creciencido de manera exponencial.
También la creación de esta información es factible de ser utilizada para promocionar su alquiler o alquileres, a las personas que dispongan de la posibilidad de alquilar departamentos, casas, etc. El presente dataset fue seleccionado en base a las posibilidades de análisis que brinda. Es interesante saber cómo se conforma el mercado de alquileres de casas en un país tan extenso como USA, donde podemos ver cómo interactúan diferentes variables en diferentes contextos. Las variables interesantes a analizar son varias en el presente dataset. A la hora de evaluar posibles personas que alquilen en el mercado de casas es importante dar información del precio, el tipo de vivienda, así como la región, el tamaño de la casa, los dormitorios, baños, acceso para discapacitados, que se está ofreciendo. Pero también es importante dar detalles sobre las posibilidades tales como poder tener gatos, perros, la cercanía a ciertos servicios para las personas, etc. Esta información es importante para los que alquilan las casas que están a la espera de maximizar sus ganancias así como empresas que generan información para luego vendersela a diferentes personas.

**Objetivo del algoritmo a realizar**

Poder predecir el precio de una casa en base a las caracteristicas que presenta para ser utilizado por las personas que tienen la posibilidad de poner inmuebles para alquilar y asi generar reditos economicos. Y tambien para aquellas personas que estan buscando un alquiler y poder predecir un precio de lo que le saldria en base a las caracteristicas buscadas.

**Algunas preguntas de interes para trabajar con el presente dataset podrian ser:**

1) ¿Existe alguna relacion entre el precio del alquiler con el tamaño en pies cuadrados?

2) ¿Hay diferencias en el precio de alquiler en base al Estado que se alquile?

3) ¿La posibilidad de tener animales aumenta el precio del alquiler?

4) ¿Cual es el tipo de alquiler predominante?

5) ¿La posibilidad de fumar en el alquiler aumenta el precio?

**Como hipotesis podemos plantear:**

1) Existe una relacion positiva entre el precio de los alquileres y el tamaño de los mismos, medido en pies cuadrados.

2) La posibilidad de tener animales, poder fumar, etc. Se relaciona de manera positiva con el precio.


### **Pasamos a ver los datos**

---


Importamos el dataset directamente desde nuestro drive

In [1]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
import numpy as np
import requests
Data1 = pd.read_csv('/content/drive/MyDrive/Entrega N1 Data Science/Casas USA/housing.csv', index_col=0)
Data1.head(5)

Mounted at /content/drive


url        region  \
id                                                                            
7049044568  https://reno.craigslist.org/apa/d/reno-beautif...  reno / tahoe   
7049047186  https://reno.craigslist.org/apa/d/reno-reduced...  reno / tahoe   
7043634882  https://reno.craigslist.org/apa/d/sparks-state...  reno / tahoe   
7049045324  https://reno.craigslist.org/apa/d/reno-1x1-fir...  reno / tahoe   
7049043759  https://reno.craigslist.org/apa/d/reno-no-long...  reno / tahoe   

                             region_url  price       type  sqfeet  beds  \
id                                                                        
7049044568  https://reno.craigslist.org   1148  apartment    1078     3   
7049047186  https://reno.craigslist.org   1200      condo    1001     2   
7043634882  https://reno.craigslist.org   1813  apartment    1683     2   
7049045324  https://reno.craigslist.org   1095  apartment     708     1   
7049043759  https://reno.craigslist.org    289  apartment     250     0   

            baths  cats_allowed  dogs_allowed  ...  wheelchair_access  \
id                                             ...                      
7049044568    2.0             1             1  ...                  0   
7049047186    2.0             0             0  ...                  0   
7043634882    2.0             1             1  ...                  0   
7049045324    1.0             1             1  ...                  0   
7049043759    1.0             1             1  ...                  1   

            electric_vehicle_charge  comes_furnished  laundry_options  \
id                                                                      
7049044568                        0                0      w/d in unit   
7049047186                        0                0      w/d hookups   
7043634882                        0                0      w/d in unit   
7049045324                        0                0      w/d in unit   
7049043759                        0                1  laundry on site   

            parking_options  \
id                            
7049044568          carport   
7049047186          carport   
7043634882  attached garage   
7049045324          carport   
7049043759              NaN   

                                                    image_url  \
id                                                              
7049044568  https://images.craigslist.org/01616_daghmBUvTC...   
7049047186  https://images.craigslist.org/00V0V_5va0MkgO9q...   
7043634882  https://images.craigslist.org/00t0t_erYqC6LgB8...   
7049045324  https://images.craigslist.org/00303_3HSJz75zlI...   
7049043759  https://images.craigslist.org/01616_fALAWFV8zQ...   

                                                  description      lat  \
id                                                                       
7049044568  Ridgeview by Vintage is where you will find al...  39.5483   
7049047186  Conveniently located in the middle town of Ren...  39.5026   
7043634882  2BD | 2BA | 1683SQFTDiscover exceptional servi...  39.6269   
7049045324  MOVE IN SPECIAL FREE WASHER/DRYER WITH 6 OR 12...  39.4477   
7049043759  Move In Today: Reno Low-Cost, Clean & Furnishe...  39.5357   

               long  state  
id                          
7049044568 -119.796     ca  
7049047186 -119.789     ca  
7043634882 -119.708     ca  
7049045324 -119.771     ca  
7049043759 -119.805     ca  

[5 rows x 21 columns]

In [2]:
df = pd.DataFrame(Data1)
df=df.drop(["url","region_url","image_url","description"],axis=1)
df.head(5)

region  price       type  sqfeet  beds  baths  cats_allowed  \
id                                                                              
7049044568  reno / tahoe   1148  apartment    1078     3    2.0             1   
7049047186  reno / tahoe   1200      condo    1001     2    2.0             0   
7043634882  reno / tahoe   1813  apartment    1683     2    2.0             1   
7049045324  reno / tahoe   1095  apartment     708     1    1.0             1   
7049043759  reno / tahoe    289  apartment     250     0    1.0             1   

            dogs_allowed  smoking_allowed  wheelchair_access  \
id                                                             
7049044568             1                0                  0   
7049047186             0                0                  0   
7043634882             1                1                  0   
7049045324             1                1                  0   
7049043759             1                1                  1   

            electric_vehicle_charge  comes_furnished  laundry_options  \
id                                                                      
7049044568                        0                0      w/d in unit   
7049047186                        0                0      w/d hookups   
7043634882                        0                0      w/d in unit   
7049045324                        0                0      w/d in unit   
7049043759                        0                1  laundry on site   

            parking_options      lat     long state  
id                                                   
7049044568          carport  39.5483 -119.796    ca  
7049047186          carport  39.5026 -119.789    ca  
7043634882  attached garage  39.6269 -119.708    ca  
7049045324          carport  39.4477 -119.771    ca  
7049043759              NaN  39.5357 -119.805    ca

In [3]:
df["baths"]=df["baths"].astype("int")
df.baths.value_counts()

1     225547
2     147811
3       6556
0       3107
4       1726
5        188
6         30
7          6
8          2
75         2
35         1
25         1
Name: baths, dtype: int64

Para no generar tanto problema podriamos eliminar los registros o rellenarlos con un valor, en este caso se selecciono el valor 0

In [4]:
df1 = df
df1 = df.fillna(0)
df1.isna().sum()


region                     0
price                      0
type                       0
sqfeet                     0
beds                       0
baths                      0
cats_allowed               0
dogs_allowed               0
smoking_allowed            0
wheelchair_access          0
electric_vehicle_charge    0
comes_furnished            0
laundry_options            0
parking_options            0
lat                        0
long                       0
state                      0
dtype: int64

In [5]:
limpieza1 = ((df["beds"]>7) | (df["baths"]>7) | (df["sqfeet"]<150) | (df["sqfeet"]>3000) | (df["price"]<100) | (df["price"]>2500))
df1 = df1[~limpieza1]

## Entrenamiento de los datos

---



In [6]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

Importamos LabelEnconder para poder clasificar las variables no numericas del dataset

In [7]:
le=LabelEncoder()
db=df1

In [8]:
db["region"]=le.fit_transform(df1["region"])
db["type"]=le.fit_transform(df1["type"])
db["state"]=le.fit_transform(df1["state"])

In [9]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 368470 entries, 7049044568 to 7049050010
Data columns (total 17 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   region                   368470 non-null  int64  
 1   price                    368470 non-null  int64  
 2   type                     368470 non-null  int64  
 3   sqfeet                   368470 non-null  int64  
 4   beds                     368470 non-null  int64  
 5   baths                    368470 non-null  int64  
 6   cats_allowed             368470 non-null  int64  
 7   dogs_allowed             368470 non-null  int64  
 8   smoking_allowed          368470 non-null  int64  
 9   wheelchair_access        368470 non-null  int64  
 10  electric_vehicle_charge  368470 non-null  int64  
 11  comes_furnished          368470 non-null  int64  
 12  laundry_options          368470 non-null  object 
 13  parking_options          368470 non-null  obje

Hay problemas porque tenemos datos que son del tipo objecto y eso no permite que se utilice LabelEconder entonces vamos a realizar transformaciones

In [10]:
db.laundry_options = db.laundry_options.replace([0], [np.nan])
db.parking_options = db.parking_options.replace([0], [np.nan])

Reemplazamos los 0 por valores NaN

In [11]:
db["laundry_options"]=le.fit_transform(df1["laundry_options"])
db["parking_options"]=le.fit_transform(df1["parking_options"])

In [12]:
db.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 368470 entries, 7049044568 to 7049050010
Data columns (total 17 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   region                   368470 non-null  int64  
 1   price                    368470 non-null  int64  
 2   type                     368470 non-null  int64  
 3   sqfeet                   368470 non-null  int64  
 4   beds                     368470 non-null  int64  
 5   baths                    368470 non-null  int64  
 6   cats_allowed             368470 non-null  int64  
 7   dogs_allowed             368470 non-null  int64  
 8   smoking_allowed          368470 non-null  int64  
 9   wheelchair_access        368470 non-null  int64  
 10  electric_vehicle_charge  368470 non-null  int64  
 11  comes_furnished          368470 non-null  int64  
 12  laundry_options          368470 non-null  int64  
 13  parking_options          368470 non-null  int6

Solucionado, ahora pasamos a realizar el entrenamiento de los datos:

In [13]:
x=db.drop(columns=["price"])
y=db["price"]

La variable a predecir sera el precio, por ende la sacamos de db que es el objeto que habiamos creado anteriormente en base al DataFrame que habiamos utilizado anteriormente en la limpieza de los datos.

In [14]:
from sklearn.preprocessing import StandardScaler
scalar=StandardScaler()
x_scaled=scalar.fit_transform(x)

Realizamos el escalado de los datos contenidos en x

In [15]:
print(x_scaled)

[[ 8.38709916e-01 -4.06773818e-01  3.03164899e-01 ...  4.05615640e-01
  -1.60886507e+00 -1.31923427e+00]
 [ 8.38709916e-01  2.95646952e-01  7.57180978e-02 ...  3.98146381e-01
  -1.60846209e+00 -1.31923427e+00]
 [ 8.38709916e-01 -4.06773818e-01  2.09024691e+00 ...  4.18462111e-01
  -1.60379914e+00 -1.31923427e+00]
 ...
 [ 8.38709916e-01 -4.06773818e-01 -3.99852486e-01 ...  4.03572626e-01
  -1.60598670e+00 -1.31923427e+00]
 [ 8.38709916e-01 -4.06773818e-01  1.87173386e-03 ...  4.07282739e-01
  -1.60351130e+00 -1.31923427e+00]
 [ 8.38709916e-01 -4.06773818e-01 -5.42515027e-02 ...  3.89173464e-01
  -1.60742588e+00 -1.31923427e+00]]


Importamos desde Scikit learn el test de entrenamiento:

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
x_train,x_test,y_train,y_test = train_test_split(x_scaled,y,test_size = 0.30,random_state=500)

In [18]:
from sklearn.linear_model import LinearRegression
lin = LinearRegression()

In [19]:
models = {"Regresion lineal" : lin}
scores= { }

In [20]:
for key, value in models.items():
    model = value
    model.fit(x_train, y_train)
    scores[key] = model.score(x_test, y_test)

In [21]:
cuadro_puntaje = pd.DataFrame(scores, index=["Puntaje de R2"]).T
cuadro_puntaje

Puntaje de R2
Regresion lineal       0.233083

Vemos que con la regresion lineal tenemos un 26% aproximadamente de precision por ende, no es nada viable utilizar este tipo de estimacion para el presente dataset, debemos buscar otro tipo de forma para entrenar los datos.

In [22]:
y_pred1=model.predict(x_test)

In [23]:
from sklearn import metrics
print('R2:',metrics.r2_score(y_test, y_pred1))
print('MSE:',metrics.mean_squared_error(y_test, y_pred1))

R2: 0.2330834317772228
MSE: 136735.24599845763


Probemos con el metodo de Bosques Aleatorios:

In [24]:
from sklearn.ensemble import RandomForestRegressor
ran = RandomForestRegressor(n_estimators=90)

In [25]:
models2 = {"Bosques Aleatorios" : ran}
scores= { }

In [26]:
for key, value in models2.items():
    model2 = value
    model2.fit(x_train, y_train)
    scores[key] = model2.score(x_test, y_test)

In [27]:
cuadro_puntaje = pd.DataFrame(scores, index=["Puntaje de R2"]).T
cuadro_puntaje

Puntaje de R2
Bosques Aleatorios       0.892982

Vemos que el resultado mejoro muchisimo con respecto a la regresion lineal, por ende vamos a evaluar con metricas este modelo de entrenamiento con bosques aleatorios

Train mayor over.

In [28]:
y_pred=ran.predict(x_test)

In [29]:
from sklearn import metrics
print('R2:',metrics.r2_score(y_test, y_pred))
print('MSE:',metrics.mean_squared_error(y_test, y_pred))

R2: 0.8929822270699447
MSE: 19080.434709747347


Vemos que el R2 muestra casi un 90% por lo tanto podriamos decir que las variables aplicadas en el metodo de bosques aleatorios explican de manera aceptable el precio de los alquileres.

In [30]:
from sklearn.model_selection import cross_validate


model = LinearRegression()
scores1 = cross_validate(model, x_train, y_train, cv=5, scoring='neg_mean_squared_error',
                        return_train_score=True)


print('Cross-Validation in Validation set MSE:', -np.mean(scores1['test_score']))

print('Cross-Validation in train set MSE:', -np.mean(scores1['train_score']))

Cross-Validation in Validation set MSE: 136808.90724163168
Cross-Validation in train set MSE: 136778.82546319702


**La diferencia entre los MSE es bastante pequeña**. Dado que la diferencia es minima, el modelo está generalizando bien no se nota overfitting (sobreajuste).

In [31]:
-np.mean(scores1['test_score']) - -np.mean(scores1['train_score'])


30.08177843465819

In [32]:
scores1

{'fit_time': array([0.16581964, 0.20690751, 0.19091129, 0.18849492, 0.212852  ]),
 'score_time': array([0.00652742, 0.00266719, 0.00263071, 0.00665188, 0.00664687]),
 'test_score': array([-136659.61699673, -135655.58083587, -137362.494193  ,
        -136780.28056982, -137586.56361274]),
 'train_score': array([-136815.1357742 , -137066.43438904, -136642.01476813,
        -136786.95112448, -136583.59126013])}

In [33]:
from sklearn.model_selection import cross_validate

scores = cross_validate(ran, x_train, y_train, cv=5, scoring='neg_mean_squared_error',
                        return_train_score=True)


print('Cross-Validation in Validation set MSE:', -np.mean(scores['test_score']))

print('Cross-Validation in train set MSE:', -np.mean(scores['train_score']))

Cross-Validation in Validation set MSE: 20885.301478837333
Cross-Validation in train set MSE: 3668.868051373511


In [34]:
-np.mean(scores['test_score']) - -np.mean(scores['train_score'])

17216.43342746382

 El **MSE : 20885** más alto en el conjunto de validación indica que el modelo no generaliza tan bien a datos nuevos. Los datos de prueba tienen un rendimiento mejor  **MSE : 3668**.

Parece que el modelo de Bosques Aleatorios esté sufriendo de **sobreajuste** a los datos de entrenamiento, lo cual implica que **ha aprendido patrones específicos de los datos de entrenamiento** en lugar de capturar patrones generales que se apliquen a nuevos datos. *Como resultado*, el modelo puede estar teniendo *dificultades para generalizar* y predecir de manera precisa los precios de las casas en datos de prueba no vistos anteriormente.